server/aesdsocket.c

```diff
@@ -11,144 +21,245 @@
+    char *buffer = calloc(BUFFER_SIZE, sizeof(char));
+    
+    while (recv(connfd, buffer, BUFFER_SIZE, 0) > 0) {
+        pthread_mutex_lock(&mutex);
+        FILE *fp = fopen("/var/tmp/aesdsocketdata", "a");
+        if (fp == NULL) {
+            perror("fopen: Failed opening file.");
+            pthread_mutex_unlock(&mutex);
+            free(buffer);
+            close(connfd);
+            return NULL;
+        } else {
+            fprintf(fp, "%s", buffer);
+            fclose(fp);
+        }
+        pthread_mutex_unlock(&mutex);
+        
+        if (strchr(buffer, '\n') != NULL) {
+            FILE *fp_out = fopen("/var/tmp/aesdsocketdata", "r");
+            if (fp_out == NULL) {
+                perror("fpopen: Failed opening file.");
+                free(buffer);
+                close(connfd);
+                return NULL;
+            } else {
+                uint32_t bytes_read = 0;
+                while ((bytes_read = fread(buffer, 1, BUFFER_SIZE, fp_out)) > 0) {
+                    send(connfd, buffer, bytes_read, 0);
+                }
+                fclose(fp_out);
+            }
+        }
+        memset(buffer, 0, BUFFER_SIZE);
+    }
+    free(buffer);
+    close(connfd);
+    return NULL;
+}
+
+void add_thread(pthread_t tid) {
+    pthread_mutex_lock(&thread_list_mutex);
+
+    node_t *new_node = (node_t *)malloc(sizeof(node_t));
+    new_node->tid = tid;
+    new_node->next = head;
+    head = new_node;
+
+    pthread_mutex_unlock(&thread_list_mutex);
+}
+
+void cleanup_threads() {
+    pthread_mutex_lock(&thread_list_mutex);
+
+    node_t *current = head;
+    while (current) {
+        pthread_join(current->tid, NULL);
+        node_t *to_free = current;
+        current = current->next;
+        free(to_free);
+    }
+
+    pthread_mutex_unlock(&thread_list_mutex);
+}
+
+
 int main(int argc, char *argv[]) {
+    // Logic to run the process as a daemon if `-d` argument is passed.
     pid_t pid, sid;
     if ((argc > 1) && (strcmp(argv[1], "-d") == 0)) {
         pid = fork();
         if (pid < 0) {
+            perror("fork: Failed to create child.");
             exit(-1);
         } else if (pid > 0) {
             exit(0);
         }
         
         sid = setsid();
         if (sid < 0) {
+            perror("setsid: Failed to acquire sid.");
             exit(-1);
         }
         
         if (chdir("/") < 0) {
+            perror("chdir: Failed to change directory.");
             exit(-1);
         }
     }
+    
+    pthread_mutex_init(&mutex, NULL);
+    
+    // Initialize syslog for logging.
     openlog("aesdsocket", LOG_CONS | LOG_PID, LOG_USER);
     
+    // Register the signal handler.
     signal(SIGINT, signal_handler);
     signal(SIGTERM, signal_handler);
+    signal(SIGALRM, signal_handler);
+
+    // Timer setup
+    timer_t timer_id;
+    struct sigevent sev;
+    memset(&sev, 0, sizeof(struct sigevent));
+    sev.sigev_notify = SIGEV_SIGNAL;
```

Comment:
@tiba6275 consider SIGEV_THREAD here so you can use the logic you are currently using in your signal handler but in a safe way.